module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,640 @@
+
+	result, err := e.results.ByBlockID(request.BlockID)
+
+	// By the time the block is sealed, the ExecutionResult must be in the db
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to lookup execution result for block: %w", err))
+	}
+
+	executionData, err := e.fetchExecutionData(ctx, result.ExecutionDataID, request.Height)
+
+	if isInvalidBlobError(err) {
+		// This means an execution result was sealed with an invalid execution data id (invalid data).
+		// Eventually, verification nodes will verify that the execution data is valid, and not sign the receipt
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("HALTING REQUESTER: invalid execution data found")
+
+		e.status.Halt()
+		e.metrics.Halted()
+		ctx.Throw(ErrRequesterHalted)
+	}
+
+	// Some or all of the blob was missing or corrupt. retry
+	if isBlobNotFoundError(err) || errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+		logger.Error().Err(err).Msg("failed to get execution data for block")
+
+		return err
+	}
+
+	// Any other error is unexpected
+	if err != nil {
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("unexpected error fetching execution data")
+
+		ctx.Throw(err)
+	}
+
+	logger.Debug().Msgf("Fetched execution data for block %d", request.Height)
+
+	request.ExecutionData = executionData
+
+	e.status.Fetched(request)
+
+	return nil
+}
+
+// fetchExecutionData fetches the ExecutionData by its ID, and times out if fetchTimeout is exceeded
+func (e *executionDataRequester) fetchExecutionData(signalerCtx irrecoverable.SignalerContext, executionDataID flow.Identifier, height uint64) (executionData *state_synchronization.ExecutionData, err error) {
+	ctx, cancel := context.WithTimeout(signalerCtx, e.config.FetchTimeout)
+	defer cancel()
+
+	start := time.Now()
+	e.metrics.ExecutionDataFetchStarted()
+	defer func() {
+		// needs to be run inside a closure so the variables are resolved when the defer is executed
+		e.metrics.ExecutionDataFetchFinished(time.Since(start), err == nil, height)
+	}()
+
+	// Get the data from the network
+	// this is a blocking call, won't be unblocked until either hitting error (including timeout) or
+	// the data is received
+	executionData, err = e.eds.Get(ctx, executionDataID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Write it to the local blobstore
+	_, _, err = e.eds.Add(signalerCtx, executionData)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to write execution data to blobstore: %w", err)
+	}
+
+	return executionData, nil
+}
+
+// Notification Worker Methods
+
+func (e *executionDataRequester) processNotificationJob(ctx irrecoverable.SignalerContext, job module.Job, complete func()) {
+	// convert job into a block entry
+	entry, err := status.JobToBlockEntry(job)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to convert job to entry: %w", err))
+	}
+
+	e.processNotification(ctx, entry)
+	complete()
```

Comment:
the caller is the job consumer. It could call complete, but I added this to allow usecases where the job may not be complete when `Run` finishes. (e.g. when a context is cancelled)

Reply:
sure. `complete()` is just a proxy for calling `consumer.NotifyJobIsDone` to mark the job done.